### PR TITLE
Fixes for static memory builds with WOLFSSL_NO_MALLOC

### PR DESCRIPTION
**Pull request by Frauschi**
`WOLFSSL_NO_MALLOC` removes only the stdio malloc fall-back, so a build that also has `--enable-staticmemory` still has a working allocator. Two places treat that configuration as having no allocator at all: one makes every RSA CA useless for chain verification, the other stops any TLS connection being established. They are independent and sit in different files. A third commit sizes the wolfCrypt test pool so `make check` exercises the configuration.

Found while running wolfCert's CI matrix, which builds it.

### 1. RSA public key never reaches the CA Signer

`ParseCert()` guarded its RSA public-key copy with `!defined(WOLFSSL_NO_MALLOC)`, while `StoreKey()` guards the equivalent copy for every non-RSA key with `!defined(WC_ASN_NO_HEAP)`. Only the latter means "genuinely no allocator", which is what the copy depends on, so a static memory build skipped it despite having somewhere to copy to.

`cert->pubKeyStored` therefore stayed 0, `FillSigner()` never populated `signer->publicKey` / `pubKeySize`, and `ConfirmSignature()` was handed a NULL key and zero size, which it rejects with `BAD_FUNC_ARG`. No certificate issued by an RSA CA could be verified against it, through `wolfSSL_CertManagerVerifyBuffer()` or TLS peer validation alike. ECC, Ed25519, Ed448 and ML-DSA CAs were unaffected, because those keys travel through `StoreKey()`.

The fix uses `WC_ASN_NO_HEAP` in all three guards in `ParseCert()`, including the one on the `char* ptr` declaration. `FreeDecodedCert()` and `FreeSigner()` already key off `pubKeyStored`, so ownership is unchanged.

### 2. Loading a static pool never set the global heap hint

An allocation made outside any `WOLFSSL_CTX` or `WOLFSSL` object can only be served from the global heap hint, and loading a pool never set it. `wolfSSL_Init()` makes exactly such an allocation: under `OPENSSL_EXTRA` it seeds the compatibility layer RNG with a NULL heap. That returned NULL, `wolfSSL_Init()` reported `WC_INIT_E`, and every `wolfSSL_CTX_new_ex()` that triggered the lazy init failed, so no TLS connection could be established at all regardless of pool size.

The two bundled examples reach this from different entry points, the server through `wolfSSL_CTX_load_static_memory()` and the client through a direct `wc_LoadStaticMemory()`, so the fix goes in `wc_LoadStaticMemory_ex()`, which both funnel through. It applies only when no hint has been set yet, so an application managing the hint itself keeps control.

Pools loaded with `WOLFMEM_IO_POOL` or `WOLFMEM_IO_POOL_FIXED` are skipped. `wc_partition_static_memory()` puts every chunk of such a pool on the heap's io list and leaves the general buckets empty, so adopting one would give a global allocator that fails every general request, and would then stop a general pool loaded afterwards from taking the hint. Load order does not matter.

`wc_UnloadStaticMemory()` now clears the hint when it refers to the heap being unloaded. The hint lives inside the caller's pool buffer, which the caller may reuse once the pool is gone, and the mutex it reaches through has just been destroyed. That part is not gated on `WOLFSSL_NO_MALLOC`: an application setting the global hint by hand has always been able to leave it dangling the same way.

Reviewers may recall 98a19f9fd, which removed commented-out `wolfSSL_SetGlobalHeapHint()` calls from both examples because the pool "does not handle all memory used on default build". That concern does not apply here: the change is confined to `WOLFSSL_NO_MALLOC`, where the alternative to serving those allocations from the pool is failing them outright. Builds with a system heap are unaffected.

### 3. wolfCrypt test pool sized for the configuration

Once the pool serves the compatibility layer's NULL-heap allocations, `gTestMemory` was too small and `openssl_test()` failed on exhaustion. The `WOLFSSL_NO_MALLOC` and `OPENSSL_EXTRA` combination now gets 1 MB; every other configuration keeps the size it had.

## Testing

With `--enable-staticmemory -DWOLFSSL_NO_MALLOC`, `make check` goes from 0 of 5 passing to 3 of 5: `resume.test`, `tls13.test` and `testsuite` all move from fail to pass. A standalone `wolfSSL_CertManagerVerifyBuffer()` on an RSA CA in that build goes from -173 to `WOLFSSL_SUCCESS`, and wolfCert's integration suite passes against it with its local workaround for defect 1 removed.

`--enable-staticmemory` without `WOLFSSL_NO_MALLOC` - the configuration the unload change newly touches - passes `make check` 5 of 5. The default configuration passes `make check` and `testwolfcrypt`, and a true no-allocator build (`-DWOLFSSL_NO_MALLOC -DNO_WOLFSSL_MEMORY`) compiles with behaviour unchanged by design. Both predicates in 1 and 2 are unchanged for those two cases, so only `WOLFSSL_NO_MALLOC` builds that do have an allocator change behaviour.

## Two failures deliberately left in place

`unit.test` reaches a structural limit rather than a sizing one: `test_wc_ChaCha20Poly1305_Encrypt_Decrypt_Big()` asks for a single 16384 byte buffer while `LARGEST_MEM_BUCKET` is 16128 for this feature set. A static pool is partitioned into fixed size buckets, so no pool size satisfies that request, and raising the ceiling changes the layout for every static memory user.

`pem.test` fails because `examples/pem/pem.c` never loads a pool at all, so its `XMALLOC(..., NULL, ...)` has no source in a build with no system heap.

Neither is a regression.

### CODE REVIEW

File: wolfcrypt/src/memory.c [hunk 1]
```diff
@@ -774,6 +774,16 @@ int wc_LoadStaticMemory_ex(WOLFSSL_HEAP_HINT** pHint,
 #endif
     *pHint = hint;
 
+#ifdef WOLFSSL_NO_MALLOC
+    /* Nothing backs a NULL-heap XMALLOC in this build, and library-global
+     * allocations have no hint to pass - the OPENSSL_EXTRA RNG that
+     * wolfSSL_Init() seeds is one, and its failure takes wolfSSL_Init() with
+     * it. Adopt the first pool loaded as the global fallback. */
+    if (wolfSSL_GetGlobalHeapHint() == NULL) {
+        (void)wolfSSL_SetGlobalHeapHint(hint);
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/memory.c]: Confirmed and fixed in 667312f02 (folded into the global-hint commit, force-pushed).

The hint lives inside the caller's pool buffer, so once the pool is unloaded the caller is free to reuse that memory, and the `memory_mutex` a later NULL-heap `XMALLOC` would lock has just been destroyed. `wc_UnloadStaticMemory()` now clears the global hint when it refers to the heap being unloaded.

I did not gate the clear on `WOLFSSL_NO_MALLOC`. The implicit assignment is new, but an application calling `wolfSSL_SetGlobalHeapHint()` by hand has always been able to leave the hint dangling across an unload, and nothing can legitimately depend on it outliving its pool. `--enable-staticmemory` without `WOLFSSL_NO_MALLOC` - the configuration that widening reaches - passes `make check` 5 of 5.

File: wolfcrypt/src/memory.c [hunk 2]
```diff
@@ -774,6 +774,16 @@ int wc_LoadStaticMemory_ex(WOLFSSL_HEAP_HINT** pHint,
 #endif
     *pHint = hint;
 
+#ifdef WOLFSSL_NO_MALLOC
+    /* Nothing backs a NULL-heap XMALLOC in this build, and library-global
+     * allocations have no hint to pass - the OPENSSL_EXTRA RNG that
+     * wolfSSL_Init() seeds is one, and its failure takes wolfSSL_Init() with
+     * it. Adopt the first pool loaded as the global fallback. */
+    if (wolfSSL_GetGlobalHeapHint() == NULL) {
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/memory.c]: Confirmed and fixed in 667312f02. This was the more serious of the two, and it went further than reported.

`wc_partition_static_memory()` places every chunk of a `WOLFMEM_IO_POOL{,_FIXED}` pool on `heap->io` and leaves `mem->ava[]` empty, so an adopted IO-only pool fails every general NULL-heap request - which is the exact `wolfSSL_Init()` failure this PR set out to fix. So the defect did not just add a new failure mode, it left the PR's headline bug live whenever an IO pool happened to load first. A standalone test that loads an IO pool, then a general one, then unloads, fails all six of its assertions against the pre-fix code, `wolfSSL_Init()` included.

Adoption is now gated on `(flag & (WOLFMEM_IO_POOL | WOLFMEM_IO_POOL_FIXED)) == 0`. Both macros are defined unconditionally in `memory.h`, so the test is correct under `WOLFSSL_STATIC_MEMORY_LEAN` too, where the flag is otherwise ignored. Load order no longer matters: an IO pool first and a general pool second now ends with the general pool holding the hint.

The PR description and the commit message both claimed a fixed IO pool could not displace the general one. That was only true when the general pool loaded first; both have been corrected.

File: wolfcrypt/src/memory.c [hunk 3]
```diff
@@ -774,6 +774,15 @@ int wc_LoadStaticMemory_ex(WOLFSSL_HEAP_HINT** pHint,
 #endif
     *pHint = hint;
 
+#ifdef WOLFSSL_NO_MALLOC
+    /* A NULL-heap XMALLOC has no other source in this build, and an IO pool
+     * holds no general buckets to serve one from. */
+    if ((wolfSSL_GetGlobalHeapHint() == NULL) &&
+            ((flag & (WOLFMEM_IO_POOL | WOLFMEM_IO_POOL_FIXED)) == 0)) {
+        (void)wolfSSL_SetGlobalHeapHint(hint);
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/memory.c]: Confirmed and fixed in ac2d35f78 (folded into the global-hint commit, force-pushed).

`wc_partition_static_memory()` returns 1 for a buffer with room for `WOLFSSL_HEAP` + `WOLFSSL_HEAP_HINT` but not for one smallest bucket - the chunking loop simply never runs - so the load reports success with every `ava[]` list empty. Against the previous head:

```
sizeof(HEAP)=256 sizeof(HINT)=40 tiny=312
tiny  load ret=0  hint=0x1049199c8 (h1=0x1049199c8)
big   load ret=0  hint=0x1049199c8 (h2=0x104919b00)
NULL-heap XMALLOC = 0x0
wolfSSL_Init = -228
```

That -228 is `WC_INIT_E`, the exact failure this PR exists to fix, so like the IO-pool finding this did not just add a new failure mode - it left the PR's headline bug live whenever a bucketless pool happened to load first.

Rather than adding a second condition I replaced the flag test with a test for a non-empty general bucket list. It subsumes the IO-pool case, since `wc_partition_static_memory()` puts every chunk of such a pool on `heap->io` and leaves `ava[]` empty, and it is more accurate under `WOLFSSL_STATIC_MEMORY_LEAN`, where the IO branch is compiled out, an IO-flagged pool does get general buckets, and the old flag test refused a perfectly usable pool. Verified in a LEAN build: an IO-flagged pool is now adopted and `wolfSSL_Init()` returns 1, where the previous head left the hint NULL and the allocation failed.

`make check` with `--enable-staticmemory --enable-opensslextra -DWOLFSSL_NO_MALLOC` is unchanged at 3 pass / 5 skip / 2 fail, the two failures being the pre-existing `pem.test` and `unit.test`, which read certs from files without ever loading a pool.

File: wolfcrypt/src/memory.c [hunk 4]
```diff
@@ -797,6 +806,12 @@ int wc_LoadStaticMemory(WOLFSSL_HEAP_HINT** pHint,
 void wc_UnloadStaticMemory(WOLFSSL_HEAP_HINT* heap)
 {
     WOLFSSL_ENTER("wc_UnloadStaticMemory");
+
+    /* the hint lives in the pool buffer, so it dies with it */
+    if (heap != NULL && wolfSSL_GetGlobalHeapHint() == (void*)heap) {
+        (void)wolfSSL_SetGlobalHeapHint(NULL);
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/memory.c:818]: Not taking this one - I do not think it is a defect.

Before this PR the global hint was only ever set by an explicit `wolfSSL_SetGlobalHeapHint()` call, so a NULL-hint allocation after unloading a pool failed exactly as it does now. Clearing the hint restores the pre-PR baseline rather than regressing past it, and the pools the finding calls stranded were just as unreachable through a NULL hint before this PR.

Promoting a survivor would need something that does not exist today: a global registry of loaded static pools, with its own locking, in a build that by definition has no allocator to build one with. The semantics would be wrong even if that were free. The application decides which pool each `WOLFSSL_CTX` or `WOLFSSL` draws from, and silently repointing the global allocator at a pool it never nominated is worse than an honest allocation failure. An application that does want a successor can name one with `wolfSSL_SetGlobalHeapHint()`, which still works: the implicit adoption only fires while the hint is NULL.

File: wolfcrypt/src/memory.c [hunk 5]
```diff
@@ -797,6 +812,12 @@ int wc_LoadStaticMemory(WOLFSSL_HEAP_HINT** pHint,
 void wc_UnloadStaticMemory(WOLFSSL_HEAP_HINT* heap)
 {
     WOLFSSL_ENTER("wc_UnloadStaticMemory");
+
+    /* the hint lives in the pool buffer, so it dies with it */
+    if (heap != NULL && wolfSSL_GetGlobalHeapHint() == (void*)heap) {
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/memory.c:817]: Not taking this one - the premise is that a concurrent `wolfSSL_SetGlobalHeapHint()` is a supported thing to do, and it never has been.

The contract sits three lines above the variable itself:

```c
/* global heap hint to fall back on when no heap hint is passed to
 * XMALLOC/XFREE
 * NOT thread safe, should be set once before any expected XMALLOC XFREE calls
 */
static void* globalHeapHint = NULL;
```

The setter is itself a non-atomic read-modify-write - it reads `globalHeapHint` into `oldHint` at :1002 and stores at :1004 - so two concurrent setters already lose an update on master, with no help from this PR. And all three allocator entry points do something sharper than a lost write: `wolfSSL_Malloc()` tests `globalHeapHint == NULL` at :1034, re-reads it at :1073 and dereferences the second read at :1078; `wolfSSL_Free()` does the same at :1222/:1256/:1261 and `wolfSSL_Realloc()` at :1369/:1385/:1390. A setter racing any of those produces a NULL dereference, not a missed assignment. If the compound access at :817-818 is a defect, those have been a more severe instance of the same defect for years, in the hot allocation path rather than in teardown.

Neither half of the suggested fix is available here. `wc_port.h` provides `WOLFSSL_ATOMIC_LOAD`/`WOLFSSL_ATOMIC_STORE` over `wolfSSL_Atomic_Int`/`wolfSSL_Atomic_Uint` only - there is no pointer-width atomic type and no compare-exchange primitive at all - and under `WOLFSSL_NO_ATOMICS` both macros degrade to a plain read and a plain write, so the CAS would collapse back into a non-atomic read-modify-write on exactly the small no-atomics targets this PR is aimed at. The locking half has no lock to hold: `memory_mutex` belongs to the pool being destroyed, the setter does not take it, and serialising on it would not exclude the racing writer.

A thread calling `wolfSSL_SetGlobalHeapHint()` concurrently with `wc_UnloadStaticMemory()` is also already in undefined behaviour four lines further down, where the same function calls `wc_FreeMutex(&heap->memory->memory_mutex)` with no synchronisation - that is #11185, and ordering the hint write correctly would not make it safe. The pool's backing buffer, which is where the hint physically lives, is the caller's to reuse the moment unload returns.

If the global hint is to become thread-safe it has to happen at the variable and cover the setter, the getter and the three allocator read sites together. A compare-exchange at the single write this PR adds would leave every pre-existing access unsynchronised while implying the variable is now safe to race.

**Comment by Frauschi** [comment]
Jenkins retest this please

**Comment by Frauschi** [comment]
Jenkins retest this please

File: wolfcrypt/src/memory.c [hunk 6]
```diff
@@ -774,6 +774,21 @@ int wc_LoadStaticMemory_ex(WOLFSSL_HEAP_HINT** pHint,
...
 
+#ifdef WOLFSSL_NO_MALLOC
+    /* a NULL-heap XMALLOC has no other source in this build, so only a pool
+     * holding general buckets can serve one */
+    if (wolfSSL_GetGlobalHeapHint() == NULL) {
+        int i;
+
+        for (i = 0; i < WOLFMEM_MAX_BUCKETS; i++) {
+            if (heap->ava[i] != NULL) {
+                (void)wolfSSL_SetGlobalHeapHint(hint);
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/memory.c]: I'm not a fan of the automatic setting and unsetting of the global heap hint. The setting of the global heap hint itself is not mutex protected so there could be obscure edge race cases here on which heap hint gets set. It's meant for an application to manage setting the global heap hint if wanted otherwise the build should fail on malloc when not set and no heap hint has been used.

Comment (JacobBarthelmeh) [reply wolfcrypt/src/memory.c]: I do like the WC_ASN_NO_HEAP fix in asn.c in this PR though!

Comment (Frauschi) [reply wolfcrypt/src/memory.c]: Fair enough - dropped it. `wolfcrypt/src/memory.c` is back to master; the applications set the hint now.

One thing: the server example couldn't set it at all. `wolfSSL_CTX_load_static_memory(&ctx, method, ...)` with a NULL ctx loads the pool and creates the CTX in one call, so `wolfSSL_Init()` runs before the caller ever sees the hint. Restructured it like the client. Any app using that form hits the same wall.

Only the standalone examples claim it, and only when unset - testsuite and unit.test run the server on a thread beside the client with function-local pools, so there the wolfCrypt test's long-lived pool is the right holder.

Also pinned the `WC_ASN_NO_HEAP` fix with a `cert_no_malloc_test()` assertion on the copy-out side, and moved the MC/DC white-box guard for that block off `WOLFSSL_NO_MALLOC`.

`--enable-staticmemory --enable-opensslextra -DWOLFSSL_NO_MALLOC` goes from 0/5 to 3 pass / 2 fail, the two remaining failing identically without these changes. Plain `--enable-staticmemory` stays 5/5.

**Review by JacobBarthelmeh — DISMISSED** [review]

File: wolfcrypt/test/test.c [hunk 7]
```diff
@@ -1529,6 +1529,11 @@ static WC_MAYBE_UNUSED Aes* test_AesGcmNew(void* heap, int declaredDevId,
         #else
             static byte gTestMemory[576*1024];  /* ML-DSA full mem */
         #endif
+    #elif defined(WOLFSSL_NO_MALLOC) && defined(OPENSSL_EXTRA) && \
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c]: Fixed, though reordering alone was not enough. At 1 MB the ML-DSA build gets past the algorithm tests and dies inside `openssl_pkey1_test()`, where the exhaustion surfaces as a NULL from `wolfSSL_X509_get_pubkey()` with no error code at all (`L=38308`). 2 MB passes.

FrodoKEM turns out to be affected the same way on its own pre-existing 1 MB arm - same failure, also cleared by 2 MB. So the arm now sits ahead of both PQ arms and asks for 2 MB when ML-DSA or FrodoKEM is enabled, 1 MB otherwise.

Both PQ configurations now run `testwolfcrypt` to completion; every other config is unchanged. Worth noting `--enable-all` does not enable ML-DSA, which is why this was not caught earlier.

File: wolfcrypt/test/test.c [hunk 8]
```diff
@@ -2354,7 +2359,7 @@ options: [-s max_relative_stack_bytes] [-m max_relative_heap_memory_bytes]\n\
         printf("unable to load static memory.\n");
         EXIT_TEST(EXIT_FAILURE);
     }
-    #ifndef OPENSSL_EXTRA
+    #if !defined(OPENSSL_EXTRA) || defined(WOLFSSL_NO_MALLOC)
     wolfSSL_SetGlobalHeapHint(HEAP_HINT);
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c]: Taken. The set is now gated on `wolfSSL_GetGlobalHeapHint() == NULL`, and the teardown only clears the hint while it is still ours.

No in-tree behaviour change - nothing installs a hint before `wolfcrypt_test()` runs, since the examples' claim is gated on `!NO_MAIN_DRIVER`. The case that matters is the one you describe: `test.c` is meant to be compiled into embedded targets as a smoke test, and under `OPENSSL_EXTRA` it never handed the hint back, so such a program lost its own pool for the rest of the process. Same first-wins rule the examples already follow.

File: wolfcrypt/test/test.c [hunk 9]
```diff
@@ -29695,6 +29700,15 @@ static wc_test_ret_t cert_no_malloc_test(void)
             ret = WC_TEST_RET_ENC_NC;
         }
     }
+#else
+    /* With an allocator the RSA key is copied out, so a Signer built from this
+     * cert keeps a public key of its own. */
+    if ((ret == 0) && ((cert.pubKeyStored != 1) ||
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c:29714]: Fixed - the assertion now carries `ParseCert()`'s whole predicate. The arm already sits under the `#else` of `WC_ASN_NO_HEAP`, so that reduces to `#elif !defined(NO_WOLFSSL_CM_VERIFY) || defined(WOLFSSL_DYN_CERT)`. RSA is the only key type involved, since `StoreKey()` and `StoreEccKey()` have no `NO_WOLFSSL_CM_VERIFY` term, which is what made the mismatch easy to miss.

For the record the configuration cannot reach the assertion today: master's own `test.c` and `asn.c` built with `-DNO_WOLFSSL_CM_VERIFY -DUSE_CERT_BUFFERS_2048` already fail at the pre-existing `FILL SIGNER test failed! error L=29773` (`fill_signer_twice_test()`, 3a6ccc13a), which depends on the same copy-out. After this change that build still stops there, unchanged, with my assertion correctly compiled out.

File: wolfcrypt/test/test.c [hunk 10]
```diff
@@ -1517,6 +1517,15 @@ static WC_MAYBE_UNUSED Aes* test_AesGcmNew(void* heap, int declaredDevId,
 #ifdef WOLFSSL_STATIC_MEMORY
     #if defined(WOLFSSL_STATIC_MEMORY_TEST_SZ)
         static byte gTestMemory[WOLFSSL_STATIC_MEMORY_TEST_SZ];
+    #elif defined(WOLFSSL_NO_MALLOC) && defined(OPENSSL_EXTRA) && \
+          !defined(BENCH_EMBEDDED)
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c]: The failure is real - I reproduced it at `OPENSSL (PKEY0)`, "error with RSA_new", `L=38065` - but I have not applied the enlarged arm to `BENCH_EMBEDDED`, because that would mean reserving 1-2 MB of `.bss` on a target that has explicitly declared itself embedded. `BENCH_EMBEDDED` is not only a benchmark knob: it gates around 40 sites in test.c that skip or shrink large tests, and settings.h auto-defines it (for example for ESP-IDF builds with benchmarks enabled), where a megabyte array would not fit.

The actual contradiction is that `openssl_pkey0_test()` is not gated on `BENCH_EMBEDDED` at all, so a "small target" build still pulls the full compatibility-layer test and asks for about 1 MB. There is no middle size that works.

So that corner now fails at build time with the override named, rather than silently reserving a megabyte or dying in `RSA_new()` at run time:

```
#error "openssl_pkey0_test() needs about 1MB of pool, far more \
than BENCH_EMBEDDED implies: set WOLFSSL_STATIC_MEMORY_TEST_SZ explicitly"
```

`WOLFSSL_STATIC_MEMORY_TEST_SZ` is still checked first, so such a target picks its own size and builds: with `-DBENCH_EMBEDDED -DWOLFSSL_STATIC_MEMORY_TEST_SZ=1048576` the whole suite passes. No CI leg combines `BENCH_EMBEDDED` with `--enable-staticmemory`, so nothing existing starts failing.

For the record this configuration is not a regression from this PR - on master it fails earlier still, at `OPENSSL test failed!` (`L=36748`), which is the bug this PR set out to fix. The main config is unchanged at 3 pass / 5 skip / 2 fail.

**Comment by Frauschi** [comment]
Jenkins retest this please

File: wolfcrypt/test/test.c [hunk 11]
```diff
@@ -1517,6 +1517,17 @@ static WC_MAYBE_UNUSED Aes* test_AesGcmNew(void* heap, int declaredDevId,
 #ifdef WOLFSSL_STATIC_MEMORY
     #if defined(WOLFSSL_STATIC_MEMORY_TEST_SZ)
         static byte gTestMemory[WOLFSSL_STATIC_MEMORY_TEST_SZ];
+    #elif defined(WOLFSSL_NO_MALLOC) && defined(OPENSSL_EXTRA)
+        /* No malloc, so the compatibility layer's NULL-heap allocations come
+         * out of this pool on top of what the algorithm tests need. */
+        #ifdef BENCH_EMBEDDED
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c:1524]: Valid, and my own doing - the #error from the previous round refuses a build that never compiles the test it names. openssl_test(), openssl_pkey0_test() and openssl_pkey1_test() are all called under `#if defined(OPENSSL_EXTRA) && !defined(WOLFCRYPT_ONLY)` (test.c:3143), and configure rejects only cryptonly + opensslall, so cryptonly + opensslextra is a supported pair that has no use for the enlarged pool at all.

Fixed as suggested: the arm is now gated on `!defined(WOLFCRYPT_ONLY)`, so a crypt-only build falls back to the ladder it used before this PR.

Verified, all with testwolfcrypt:

- `--enable-cryptonly --enable-opensslextra --enable-staticmemory -DWOLFSSL_NO_MALLOC -DBENCH_EMBEDDED` reproduces the #error before the change; after it, builds and exits 0 on the 14000-byte arm.
- The same configuration without BENCH_EMBEDDED now takes the 160000-byte SP-math arm instead of 1 MB and exits 0.
- The non-crypt-only `--enable-opensslextra --enable-staticmemory -DWOLFSSL_NO_MALLOC` build is untouched and still exits 0.

The workflow also had no coverage for any of this - all five no-malloc entries omit --enable-opensslextra - so I added two: opensslextra-staticmemory for the configuration this PR targets, and cryptonly-opensslextra-embedded so that a regression of the exclusion above fails the build rather than silently reserving a megabyte.

File: wolfcrypt/test/test.c [hunk 12]
```diff
@@ -1517,6 +1517,18 @@ static WC_MAYBE_UNUSED Aes* test_AesGcmNew(void* heap, int declaredDevId,
 #ifdef WOLFSSL_STATIC_MEMORY
     #if defined(WOLFSSL_STATIC_MEMORY_TEST_SZ)
         static byte gTestMemory[WOLFSSL_STATIC_MEMORY_TEST_SZ];
+    #elif defined(WOLFSSL_NO_MALLOC) && defined(OPENSSL_EXTRA) && \
+          !defined(WOLFCRYPT_ONLY)
+        /* No malloc, so the compatibility layer's NULL-heap allocations come
+         * out of this pool on top of what the algorithm tests need. */
+        #ifdef BENCH_EMBEDDED
+            #error "openssl_pkey0_test() needs about 1MB of pool, far more \
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c:1525]: Valid, and I had to measure against my own objection on this one. My first read was that it could not be narrowed: openssl_test() is not gated on NO_RSA at all, so I expected the 14000-byte arm to trade the #error for a silent pool exhaustion. It does not - openssl_test()'s large allocations are all on the RSA and EVP_PKEY paths.

Measured with --disable-rsa --enable-opensslextra --enable-staticmemory -DWOLFSSL_NO_MALLOC:

- with -DBENCH_EMBEDDED: takes the 14000-byte arm, whole suite passes, testwolfcrypt exits 0.
- without it: takes the 160000-byte SP-math arm instead of 1 MB, exits 0.

Fixed by excluding the whole arm on `!defined(NO_RSA) && !defined(NO_SHA)` rather than only the rejection. Gating just the #error would drop BENCH_EMBEDDED + NO_RSA into the 1 MB #else, which is the megabyte the #error exists to refuse, and would leave the plain NO_RSA build reserving it as well. Builds with RSA are unchanged: 1 MB, or 2 MB with ML-DSA or FrodoKEM.

File: wolfcrypt/test/test.c [hunk 13]
```diff
@@ -29695,6 +29709,15 @@ static wc_test_ret_t cert_no_malloc_test(void)
             ret = WC_TEST_RET_ENC_NC;
         }
     }
+#elif !defined(NO_WOLFSSL_CM_VERIFY) || defined(WOLFSSL_DYN_CERT)
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/test/test.c:29712]: Valid, and it reaches further than the assertion. No configure option defines WOLFSSL_TEST_CERT or USE_CERT_BUFFERS_2048, so in that entry cert_no_malloc_test() and fill_signer_twice_test() were both compiled out, and nothing in CI touched the copy-out this series restores.

Added -DWOLFSSL_TEST_CERT -DUSE_CERT_BUFFERS_2048 -DUSE_CERT_BUFFERS_256 to opensslextra-staticmemory. Verified with the entry's exact flags: the build takes the allocator arm (WC_ASN_NO_HEAP undefined, NO_WOLFSSL_CM_VERIFY unset), and both "FILL SIGNER test passed!" and "CERT NOMALLOC test passed!" now appear where neither test ran before.

Also confirmed the entry catches a regression rather than merely running the test: putting ParseCert()'s public-key guard back to WOLFSSL_NO_MALLOC in the same tree makes testwolfcrypt exit 1, at fill_signer_twice_test() first, with the mirror assertion behind it.

**Review by JacobBarthelmeh — APPROVED** [review]
